src/background_music/Integrator.h

```diff
@@ -0,0 +1,222 @@
+        return max;
+    }
+
+    /**
+     * Returns the history multiplier for integration over the provided number
+     * of samples, that can be a fraction. If the fraction is smaller than
+     * {@code minimum_samples}, zero is returned. If the number is bigger than
+     * {@code maximum_samples} the history multiplier for {@code
+     * maximum_samples} is returned.
+     * @param samples The number of samples
+     * @return the history multiplier that lies between [0 .. 1>.
+     */
+    static inline S calculate_history_multiplier_from_samples(S samples)
+    {
+        S absolute = fabs(samples);
+        return absolute < minimum_samples()   ? 0
+               : absolute > maximum_samples() ? exp(-1 / maximum_samples())
+                                            : exp(-1.0 / absolute);
+    }
+
+    /**
+     * Calculates the history multiplier from a given input multiplier and vice
+     * versa.
+     * @param multiplier The multiplier
+     * @return The other multiplier
+     */
+    [[maybe_unused]] static inline S calculate_other_multiplier(S multiplier)
+    {
+        return 1 - multiplier;
+    }
+
+    /**
+     * Calculates the number of samples for the provided history multiplier. The
+     * result is always positive or zero, but not always a whole number. If the
+     * history multiplier is not valid, std::invalid_argument is thrown.
+     * The choice was made not to return -1 on error, as that easily leads to
+     * segfaults.
+     */
+    [[maybe_unused]] static inline S
+    calculate_samples_from_history_multiplier(S history_multiplier)
+    {
+        double absolute = fabs(history_multiplier);
+        if (absolute < std::numeric_limits<S>::epsilon()())
+        {
+            return 0;
+        }
+        else if (absolute <= (1 - std::numeric_limits<S>::epsilon()()))
+        {
+            return 1.0 / log(absolute);
+        }
+        throw std::invalid_argument(
+            "IntegratorBase::calculate_samples_from_history_multiplier("
+            "multiplier): multiplier out of range.");
+    }
+
+    /**
+     * By default, the integrator does not integrate and just passes-through the
+     * input.
+     */
+    inline constexpr Integrator() : history_multiplier_(0), input_multiplier_(1) {}
+
+    /**
+     * Creates an integrator over the provided number of samples.
+     * @param samples The number of samples
+     */
+    [[maybe_unused]] inline explicit Integrator(S samples)
+        : history_multiplier_(
+              calculate_history_multiplier_from_samples(samples)),
+          input_multiplier_(1.0 - history_multiplier_)
+    {
+    }
+
+    /**
+     * Creates an integrator of the the provided number of samples that is
+     * calculated from the time and sample rate.
+     * @param time The number of time units
+     * @param sample_rate The sample-rate samples per unit of time
+     */
+    [[maybe_unused]] inline Integrator(S time, S sample_rate)
```

Comment:
This is a self-contained plugin, not a general-purpose shared library. If there is unused code, please delete it.